### PR TITLE
ui: easy links from one node's to another one's

### LINKED ISSUE #31564: ui: easy links from one node's admin UI to another one's
For some debug pages, you want to be on a specific node (e.g. logspy). It'd be really cool if the admin UI would let me jump from a node to another, or at least if it printed (all?) the IPs of all the node's so that I can construct the URL by hand.

Particularly with roachprod clusters, it's not trivial to go from a node id to an IP - the order of nodes does not generally correspond to the order of machines in the roachprod cluster.
Anything you can give me here, I'll take.

### PR DESCRIPTION
**Pull request by vladlos**
added select box with nodes addresses

Resolves: #31564

Release note (ui): none

### CODE REVIEW

**Review by koorosh — CHANGES_REQUESTED** [review]
@elkmaster , 
- Implementation of nodes selectors is coupled with Debug page. I suggest to refactor it in
separate connected component which will incorporate all logic and representation.
As result you won't need to make Debug page connected.
- Dropdown component which is used here had to be restyled for `blue` theme. Instead you might use
available Dropdown component which is already styled according to new designs (`src/components/dropdown`).

**Review by dhartunian — CHANGES_REQUESTED** [review]
@elkmaster this doesn't run as-is. I get a console error. Can you take a look?

I'm testing with `cockroach demo movr --nodes 9 --global --with-load --geo-partitioned-replicas`

**Comment by elkmaster** [comment]
@dhartunian, this is some backend issue with demo command. It not works for me either in latest master: `Error: unknown flag: --nodes
Failed running "demo"`

**Comment by dhartunian** [comment]
@elkmaster are you running latest cockroach? I can run the command on `19.2.2` installed via brew.

The latest build had a ui test fail:
```
[Test web UI] FAILED TESTS:
[Test web UI]     '/debug' path
[Test web UI]       ✖ routes to <Debug> component
[Test web UI]         HeadlessChrome 79.0.3945 (Linux 0.0.0)
[Test web UI]       AssertionError: expected { length: 0 } to have a length of 1 but got 0
[Test web UI]           at Context.<anonymous> (src/app.spec.tsx:294229:51)
[Test web UI] 
```

**Comment by elkmaster** [comment]
never mind, I run command incorrectly. fixed tests, should be ok now

File: pkg/ui/src/views/reports/containers/debug/index.tsx [hunk 1]
```diff
@@ -79,285 +80,399 @@ function DebugPanelLink(props: { name: string, url: string,  note: string }) {
...
+            <p>
+              The following pages are meant for advanced monitoring and troubleshooting.
+              Note that these pages are experimental and undocumented. If you find an issue,
+              let us know through{" "}
+              <a href={ COMMUNITY_URL }>these channels.</a>
+            </p>
+          </InfoBox>
+          <div className="debug-header__annotations">
+            <LicenseType />
+            <Dropdown
```
Comment (dhartunian) [inline pkg/ui/src/views/reports/containers/debug/index.tsx]: This file is already very long, please extract this into a separate component in its own file.

File: pkg/ui/src/views/reports/containers/debug/index.tsx [hunk 2]
```diff
@@ -79,285 +80,399 @@ function DebugPanelLink(props: { name: string, url: string,  note: string }) {
...
+}
+
+interface IDebugState {
+  selected: string;
+  location: string;
+  opened: boolean;
+  width: number;
+}
+
+const pattern = /[-a-zA-Z0-9@:%._\+~#=]{1,256}\.[a-zA-Z0-9()]{1,6}\b([-a-zA-Z0-9()@%_\+.~#?&=]*)?/;
```
Comment (dhartunian) [inline pkg/ui/src/views/reports/containers/debug/index.tsx]: Can we do this without regexes please? I can't read them. If you're going to use one, please name it after what it matches and explain why it's necessary.

File: pkg/ui/src/views/reports/containers/debug/index.tsx [hunk 3]
```diff
@@ -79,285 +80,399 @@ function DebugPanelLink(props: { name: string, url: string,  note: string }) {
...
+                      </span>
+                      <div className="Select-control">
+                        <div className="Select-arrow-zone">
+                          {this.arrowRenderer(opened)}
+                        </div>
+                      </div>
+                    </div>
+                    {opened && (
+                      <div className={`dropdown-options`} style={{ left }}>
+                        {this.renderOptions()}
```
Comment (dhartunian) [inline pkg/ui/src/views/reports/containers/debug/index.tsx]: Avoid using `renderX()` functions. Extract into a pure component.

File: pkg/ui/src/views/reports/containers/debug/index.tsx [hunk 4]
```diff
@@ -79,285 +80,399 @@ function DebugPanelLink(props: { name: string, url: string,  note: string }) {
...
+              Note that these pages are experimental and undocumented. If you find an issue,
+              let us know through{" "}
+              <a href={ COMMUNITY_URL }>these channels.</a>
+            </p>
+          </InfoBox>
+          <div className="debug-header__annotations">
+            <LicenseType />
+            <Dropdown
+              title="Web server"
+              options={[]}
```
Comment (dhartunian) [inline pkg/ui/src/views/reports/containers/debug/index.tsx]: Why doesn't the dropdown ever have options?

Comment (dhartunian) [inline pkg/ui/src/views/reports/containers/debug/index.tsx]: Why doesn't the dropdown ever have options?

File: pkg/ui/src/views/shared/components/dropdown/dropdown.styl [hunk 5]
```diff
@@ -127,4 +127,67 @@ $dropdown-hover-color = darken($background-color, 2.5%)
   .Select-menu-outer, .Select-menu
     max-height 450px
 
+.dropdown__palette--blue
```
Comment (dhartunian) [inline pkg/ui/src/views/shared/components/dropdown/dropdown.styl]: Why did we need a new dropdown palette?

File: pkg/ui/src/views/shared/components/dropdown/index.tsx [hunk 6]
```diff
@@ -79,12 +87,14 @@ export default class Dropdown extends React.Component<DropdownOwnProps, {}> {
   }
 
   render() {
-    const { selected, options, onChange, onArrowClick, disabledArrows, content, isTimeRange } = this.props;
+    const { selected, options, onChange, onArrowClick, disabledArrows, content, isTimeRange, palette, placeholder, focused } = this.props;
 
     const className = classNames(
       "dropdown",
+      `dropdown__palette--${palette}`,
```
Comment (dhartunian) [inline pkg/ui/src/views/shared/components/dropdown/index.tsx]: I'd like to avoid making this component more and more configurable. Ideally the API surface area and the className surface area becomes smaller over time.

File: pkg/ui/src/views/shared/components/dropdown/index.tsx [hunk 7]
```diff
@@ -117,6 +127,7 @@ export default class Dropdown extends React.Component<DropdownOwnProps, {}> {
         value={selected}
         onChange={onChange}
         ref={this.selectRef}
+        placeholder={placeholder}
```
Comment (dhartunian) [inline pkg/ui/src/views/shared/components/dropdown/index.tsx]: Why do we need this?

Comment (dhartunian) [inline pkg/ui/src/views/shared/components/dropdown/index.tsx]: Why do we need this?

File: pkg/ui/src/views/reports/containers/debug/index.tsx [hunk 8]
```diff
@@ -79,285 +80,399 @@ function DebugPanelLink(props: { name: string, url: string,  note: string }) {
...
-        </InfoBox>
+interface IDebugProps {
+  nodesSummary: NodesSummary;
+  dataSource: DecommissionedNodeStatusRow[];
+}
+
+interface IDebugState {
+  selected: string;
+  location: string;
+  opened: boolean;
```
Comment (dhartunian) [inline pkg/ui/src/views/reports/containers/debug/index.tsx]: State like this should live in the extracted dropdown component. It doesn't make sense for `Debug` to have an `opened` state since the open refers to the dropdown, not debug.

File: pkg/ui/src/views/reports/containers/debug/index.tsx [hunk 9]
```diff
@@ -79,285 +80,399 @@ function DebugPanelLink(props: { name: string, url: string,  note: string }) {
...
+  }
+
+  renderOptions = () => {
+    const { selected } = this.state;
+    return this.getOptionsForSelect().map((option: DropdownOption) => (
+      <a onClick={this.selectApp(option)} className={`option ${selected === option.value ? "selected" : ""}`}>{option.label}</a>
+    ));
+  }
+
+  updateDimensions = () => {
```
Comment (dhartunian) [inline pkg/ui/src/views/reports/containers/debug/index.tsx]: Why is this necessary? Seems like something CSS should solve.

File: pkg/ui/src/views/reports/containers/debug/index.tsx [hunk 10]
```diff
@@ -79,285 +80,399 @@ function DebugPanelLink(props: { name: string, url: string,  note: string }) {
...
+    this.setState({
+      width: window.innerWidth,
+    });
+  }
+
+  render() {
+    const { selected, width, opened } = this.state;
+    const containerLeft = this.rangeContainer.current ? this.rangeContainer.current.getBoundingClientRect().left : 0;
```
Comment (dhartunian) [inline pkg/ui/src/views/reports/containers/debug/index.tsx]: Can we avoid using APIs like this? I'd rather solve the root cause that's preventing us from using CSS instead of working around existing technical debt.

File: pkg/ui/src/views/reports/containers/debug/index.tsx [hunk 11]
```diff
@@ -8,22 +8,23 @@
 // by the Apache License, Version 2.0, included in the file
 // licenses/APL.txt.
 
+import { Icon } from "antd";
```
Comment (dhartunian) [inline pkg/ui/src/views/reports/containers/debug/index.tsx]: Please remove import reordering from the PR

**Review by dhartunian — CHANGES_REQUESTED** [review]

**Comment by dhartunian** [comment]
@elkmaster Thanks for addressing all the comments! Appreciate the code reorganization, much easier to follow.

I've decided to make backend changes to ease the reporting of addresses on the frontend.

I'm working on adding `http_address` to the `nodes.data.desc` data object. Let's put this PR on hold until that's in there so you can cleanly link to those URLs without doing any parsing of unrelated pieces of data.
